Upserting input data based on personas (#5)
ttsim-dev/gettsim-personas: README.md

# GETTSIM Personas

This repository provides example personas to use as input data for GETTSIM. Personas
depict typical households, which GETTSIM's users might be interested in.

## Basic Concepts

### Persona Attributes

Each persona has the following attributes:

> - `name`: the name of the persona

Reviewer: Unique identifier of the persona

> - `description`: a description of the persona constellation

Reviewer: Should be clear from the docstring what is the idea behind description, purpose — not obvious as is, and what is the difference

# GETTSIM Personas

This repository provides example personas to use as input data for GETTSIM. Personas

> depict typical household constellation GETTSIM's users might be interested in.

Reviewer: ```suggestion
depict typical households, which GETTSIM's users might be interested in.
```

- `name`: the name of the persona
- `description`: a description of the persona constellation
- `purpose`: the purpose of this persona, i.e. what exactly it is trying to depict
- `start_date`: (Optional) the date from which the persona is valid
- `end_date`: (Optional) the date until which the persona is valid
- `policy_inputs`: the policy inputs that are used to calculate the targets
- `policy_inputs_overriding_functions`: a dictionary of functions that are used to
  override GETTSIM's policy functions
- `targets_tree`: the targets that can be computed for this persona

> [!WARNING]
> Be careful when using personas in a different context than intended. Many personas
> overwrite GETTSIM's policy functions via `policy_inputs_overriding_functions` or
> represent specific household structures. Always check whether a persona is suitable
> for your use case.

## Usage

The following example shows how to use a persona to calculate taxes and transfers for a
household.

### Load personas for a specific date

As personas are date-specific, we first need to load the personas for the date we are
interested in.

```python
from gettsim_personas import personas_for_date

> personas = personas_for_date("2025-01-01")

Reviewer: ```suggestion
jan_01_2025 = "2025-01-01"
personas = get_personas(date=jan_01_2025)
```

The `_for_date` part is redundant given that we have a required keyword argument there.

Define a date (feel free to use generic `date` or whatever if you prefer) to make clear that this better matches the date used for the policy environment.

> couple_1_child = personas.couple_1_child

Reviewer: ```suggestion
couple_1_child = personas.couple_1_child_no_means_tested_transfers
```
That is a crucial feature of this persona, it seems? Seems a bit inconsistent to ignore Wohngeld but not Kinderzuschlag, btw.

```

The `personas` objects contains all available personas for the given date. We selected
the `couple_1_child` persona to compute taxes and transfers.

### Compute taxes and transfers

Now, we can compute taxes and transfers for the selected persona.

```python
from gettsim import set_up_policy_environment, compute_taxes_and_transfers

> environment = set_up_policy_environment(date="2025-01-01")

Reviewer: ```suggestion
environment = set_up_policy_environment(date=jan_01_2025)
```

Doesn't `oss` work ?

Reviewer: Does not matter at the end of the day, I think. DataFrame + mapping will be the way most people will use GETTSIM, but both ways will feel very natural to any one who invests more than a day into it.

result = compute_taxes_and_transfers(
    data_tree=couple_1_child.input_data,
    environment=environment,
    targets_tree=couple_1_child.targets_tree,
)
```

> ### Upserting input data

Reviewer: ```suggestion
### Updating / inserting input data
```

Please spell out what happens below. I understand what is going on after thinking about it for a while, but it will be very opaque to the non-initiated. Maybe something like this:
1. Display the current situation (what does `couple_1_child.couple_1_child_no_means_tested_transfers` contain, at least in the column to be modified)
2. State what is the goal ("We want to calculate taxes for two different income levels of the husband. Earnings of the wife and of the kid are zero in both cases")
3. Explain how `"bruttolohn_m"` below is broadcasted.
4. The warning then makes much more sense, now that we know the logic of how broadcasting works.

Then add an example where we do the same based on a linspace. Maybe better for the same couple including transfers, though?

Users may modify the persona by providing their own input data. Users can either add new
inputs or modify existing inputs. The function `upsert_input_data` adds the
user-provided input data to the persona's input data and broadcasts the persona's data
to match the length of the user-provided data.

```python
from gettsim_personas import upsert_input_data

bruttolohn_to_upsert = {
    "einkommensteuer": {
        "einkünfte": {
            "aus_nichtselbstständiger_arbeit": {
                "bruttolohn_m": pd.Series([4000.0, 0.0, 0.0, 6000.0, 0.0, 0.0])
            }
        }
    }
}

> upserted_input_data = upsert_input_data(

Reviewer: ```suggestion
input_data_with_2_income_levels = upsert_input_data(
```

Then re-do calculations of targets.
